[run-webkit-tests] Add visionOS

JonWBedard opened the pull request:
#### 3159d8008436cd8e8d110d3d5117e954371b259b
<pre>
[run-webkit-tests] Add visionOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=268567">https://bugs.webkit.org/show_bug.cgi?id=268567</a>
<a href="https://rdar.apple.com/112615351">rdar://112615351</a>

Reviewed by Ryan Haddad.

* Tools/Scripts/webkitpy/common/version_name_map.py:
(VersionNameMap.__init__): Add visionOS.
* Tools/Scripts/webkitpy/port/factory.py:
(platform_options): Add --visionos-simulator flag.
(PortFactory): Add visionOS port.
* Tools/Scripts/webkitpy/port/visionos.py: Added.
(VisionOSPort): Added.
* Tools/Scripts/webkitpy/port/visionos_simulator.py: Added.
(VisionOSSimulatorPort): Added.
* Tools/Scripts/webkitpy/port/visionos_simulator_unittest.py: Added.
(VisionOSSimulatorTest): Added.
* Tools/Scripts/webkitpy/port/visionos_testcase.py: Added.
(VisionOSTest): Added.
* Tools/Scripts/webkitpy/xcode/device_type.py:
(DeviceType._define_software_variant_from_hardware_family): Handle Vision Pro.
(DeviceType.standardize_hardware_type): Strip &apos;Apple&apos; from &apos;Apple Vision Pro&apos;
to match naming conventions of other simulated hardware.
* Tools/Scripts/webkitpy/xcode/device_type_unittest.py:
(DeviceTypeTest.test_visionos): Added.

Canonical link: <a href="https://commits.webkit.org/274484@main">https://commits.webkit.org/274484@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4caf38d50a8f0f4d0431cf54fcdb47fbdc8c8a9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39250 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18229 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41603 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41784 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41556 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21085 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15558 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39824 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/15349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13330 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/39115 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/34962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43062 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/35644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/11608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/15668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8780 "Build was cancelled. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Checked out pull request; Reviewed by Ryan Haddad; Compiled WebKit (cancelled)") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/15331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5123 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/webkitpy/common/version_name_map.py:
```diff
@@ -70,6 +70,7 @@ def __init__(self, platform=None):
             'ios': self._automap_to_major_version('iOS', minimum=Version(10), maximum=Version(17)),
             'tvos': self._automap_to_major_version('tvOS', minimum=Version(10), maximum=Version(17)),
             'watchos': self._automap_to_major_version('watchOS', minimum=Version(1), maximum=Version(10)),
+            'visionos': {'visionOS 1': Version(1)},
```
I know this is redundant here, but can we be consistent with the above and just use:
```suggestion
            'visionos': self._automap_to_major_version('visionOS', minimum=Version(1), maximum=Version(1)),
```

JonWBedard replied:
We actually cannot because `self._automap_to_major_version` doesn't work if the two versions match.

gsnedders replied:
https://github.com/WebKit/WebKit/pull/23775 suggests otherwise 🙃

gsnedders commented on Tools/Scripts/webkitpy/port/visionos.py:
```diff
@@ -0,0 +1,97 @@
+            return None
+        return VersionNameMap.map(self.host.platform).to_name(self._os_version, platform=VisionOSPort.port_name)
+
+    def default_baseline_search_path(self, **kwargs):
```
Nit: can we not move this to somewhere shared between all the Apple embedded ports? I guess this is slightly complicated because we need to be able to add multiple platform names (`VisionOSPort.port_name` and `self.port_name`)?

JonWBedard replied:
The issue is that while watchOS and visionOS share a baseline search path format, iOS and iPadOS don't. All of this is made further complicated by the fact that we're probably going to end up adding a UIKit baseline in short order here.

gsnedders replied:
iPadOS is the only real special-case here, right?

gsnedders commented on Tools/Scripts/webkitpy/port/visionos.py:
```diff
@@ -0,0 +1,97 @@
+        for version in versions_to_fallback:
+            if apple_additions():
+                apple_name = VersionNameMap.map(self.host.platform).to_name(version, platform=VisionOSPort.port_name, table=INTERNAL_TABLE)
+                expectations.append(self._apple_baseline_path('{}-{}'.format(self.port_name, apple_name.lower().replace(' ', ''))))
```
This assumes that `VisionOSPort` will never be directly initiated, right? (And any potential on-device testing would be a `visionos-device` port, if we follow iOS naming?)

JonWBedard replied:
Yes, the `<platform>Port` classes are never directly initiated.

gsnedders commented on Tools/Scripts/webkitpy/port/visionos.py:
```diff
@@ -0,0 +1,97 @@
+        super(VisionOSPort, self).__init__(*args, **kwargs)
+
+        if not self.get_option('webkit_test_runner', False):
+            raise RuntimeError('DumpRenderTree is not supported on this platform.')
```
Nit, given this is more copy/paste code…

`RuntimeError` doesn't really seem appropriate here; it's more a `ValueError`? I do wonder if a `UnsupportedConfigurationError` would be worthwhile adding as a subclass, because we also have other cases like this.

gsnedders commented on Tools/Scripts/webkitpy/port/visionos_simulator_unittest.py:
```diff
@@ -0,0 +1,82 @@
+
+
+class VisionOSSimulatorTest(visionos_testcase.VisionOSTest):
+    # FIXME: https://bugs.webkit.org/show_bug.cgi?id=173107
```
A FIXME pointing at a RESOLVED WONTFIX bug is nonsense.

gsnedders commented on Tools/Scripts/webkitpy/port/visionos_simulator_unittest.py:
```diff
@@ -0,0 +1,82 @@
+        port._run_script = run_script
+        self.assertEqual(port.architecture(), 'arm64')
+        port._build_driver()
+        self.assertEqual(self.args, ['--sdk', 'xrsimulator', 'ARCHS=arm64'])
```
How is this in any way 32-bit?

gsnedders reviewed (commented):
[non-reviewer]

I realise a lot of this is copy/paste code, but we now have _so many_ copies of all the iOS/watchOS/visionOS code that it feels like we really should try and get rid of all this duplication.

I'm sure some of these comments also apply to iOS/watchOS ports, but I've tried to explicitly label them as nits where that's likely the case.

ryanhaddad reviewed (approved):
r=me

JonWBedard committed:
[run-webkit-tests] Add visionOS
https://bugs.webkit.org/show_bug.cgi?id=268567
rdar://112615351

Reviewed by Ryan Haddad.

* Tools/Scripts/webkitpy/common/version_name_map.py:
(VersionNameMap.__init__): Add visionOS.
* Tools/Scripts/webkitpy/port/factory.py:
(platform_options): Add --visionos-simulator flag.
(PortFactory): Add visionOS port.
* Tools/Scripts/webkitpy/port/visionos.py: Added.
(VisionOSPort): Added.
* Tools/Scripts/webkitpy/port/visionos_simulator.py: Added.
(VisionOSSimulatorPort): Added.
* Tools/Scripts/webkitpy/port/visionos_simulator_unittest.py: Added.
(VisionOSSimulatorTest): Added.
* Tools/Scripts/webkitpy/port/visionos_testcase.py: Added.
(VisionOSTest): Added.
* Tools/Scripts/webkitpy/xcode/device_type.py:
(DeviceType._define_software_variant_from_hardware_family): Handle Vision Pro.
(DeviceType.standardize_hardware_type): Strip 'Apple' from 'Apple Vision Pro'
to match naming conventions of other simulated hardware.
* Tools/Scripts/webkitpy/xcode/device_type_unittest.py:
(DeviceTypeTest.test_visionos): Added.

Canonical link: https://commits.webkit.org/274484@main
